In [1]:
import pandas as pd

In [2]:
import numpy as np

# EDA

In [3]:
database=pd.read_excel("Bank_Personal_Loan_Modelling.xlsx",sheet_name=1)

In [4]:
database

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  \
0        1   25           1      49     91107       4    1.6          1   
1        2   45          19      34     90089       3    1.5          1   
2        3   39          15      11     94720       1    1.0          1   
3        4   35           9     100     94112       1    2.7          2   
4        5   35           8      45     91330       4    1.0          2   
...    ...  ...         ...     ...       ...     ...    ...        ...   
4995  4996   29           3      40     92697       1    1.9          3   
4996  4997   30           4      15     92037       4    0.4          1   
4997  4998   63          39      24     93023       2    0.3          3   
4998  4999   65          40      49     90034       3    0.5          2   
4999  5000   28           4      83     92612       3    0.8          1   

      Mortgage  Personal Loan  Securities Account  CD Account  Online  \
0            0              0                   1           0       0   
1            0              0                   1           0       0   
2            0              0                   0           0       0   
3            0              0                   0           0       0   
4            0              0                   0           0       0   
...        ...            ...                 ...         ...     ...   
4995         0              0                   0           0       1   
4996        85              0                   0           0       1   
4997         0              0                   0           0       0   
4998         0              0                   0           0       1   
4999         0              0                   0           0       1   

      CreditCard  
0              0  
1              0  
2              0  
3              0  
4              1  
...          ...  
4995           0  
4996           0  
4997           0  
4998           0  
4999           1  

[5000 rows x 14 columns]

In [5]:
database.describe().T

count          mean          std     min       25%  \
ID                  5000.0   2500.500000  1443.520003     1.0   1250.75   
Age                 5000.0     45.338400    11.463166    23.0     35.00   
Experience          5000.0     20.104600    11.467954    -3.0     10.00   
Income              5000.0     73.774200    46.033729     8.0     39.00   
ZIP Code            5000.0  93152.503000  2121.852197  9307.0  91911.00   
Family              5000.0      2.396400     1.147663     1.0      1.00   
CCAvg               5000.0      1.937913     1.747666     0.0      0.70   
Education           5000.0      1.881000     0.839869     1.0      1.00   
Mortgage            5000.0     56.498800   101.713802     0.0      0.00   
Personal Loan       5000.0      0.096000     0.294621     0.0      0.00   
Securities Account  5000.0      0.104400     0.305809     0.0      0.00   
CD Account          5000.0      0.060400     0.238250     0.0      0.00   
Online              5000.0      0.596800     0.490589     0.0      0.00   
CreditCard          5000.0      0.294000     0.455637     0.0      0.00   

                        50%       75%      max  
ID                   2500.5   3750.25   5000.0  
Age                    45.0     55.00     67.0  
Experience             20.0     30.00     43.0  
Income                 64.0     98.00    224.0  
ZIP Code            93437.0  94608.00  96651.0  
Family                  2.0      3.00      4.0  
CCAvg                   1.5      2.50     10.0  
Education               2.0      3.00      3.0  
Mortgage                0.0    101.00    635.0  
Personal Loan           0.0      0.00      1.0  
Securities Account      0.0      0.00      1.0  
CD Account              0.0      0.00      1.0  
Online                  1.0      1.00      1.0  
CreditCard              0.0      1.00      1.0

In [6]:
database.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  5000 non-null   int64  
 1   Age                 5000 non-null   int64  
 2   Experience          5000 non-null   int64  
 3   Income              5000 non-null   int64  
 4   ZIP Code            5000 non-null   int64  
 5   Family              5000 non-null   int64  
 6   CCAvg               5000 non-null   float64
 7   Education           5000 non-null   int64  
 8   Mortgage            5000 non-null   int64  
 9   Personal Loan       5000 non-null   int64  
 10  Securities Account  5000 non-null   int64  
 11  CD Account          5000 non-null   int64  
 12  Online              5000 non-null   int64  
 13  CreditCard          5000 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 547.0 KB


In [7]:
database.columns

Index(['ID', 'Age', 'Experience', 'Income', 'ZIP Code', 'Family', 'CCAvg',
       'Education', 'Mortgage', 'Personal Loan', 'Securities Account',
       'CD Account', 'Online', 'CreditCard'],
      dtype='object')

In [8]:
database.dtypes

ID                      int64
Age                     int64
Experience              int64
Income                  int64
ZIP Code                int64
Family                  int64
CCAvg                 float64
Education               int64
Mortgage                int64
Personal Loan           int64
Securities Account      int64
CD Account              int64
Online                  int64
CreditCard              int64
dtype: object

In [9]:
database.isnull() is True

False

In [10]:
database.corr()

ID       Age  Experience    Income  ZIP Code  \
ID                  1.000000 -0.008473   -0.008326 -0.017695  0.013432   
Age                -0.008473  1.000000    0.994215 -0.055269 -0.029216   
Experience         -0.008326  0.994215    1.000000 -0.046574 -0.028626   
Income             -0.017695 -0.055269   -0.046574  1.000000 -0.016410   
ZIP Code            0.013432 -0.029216   -0.028626 -0.016410  1.000000   
Family             -0.016797 -0.046418   -0.052563 -0.157501  0.011778   
CCAvg              -0.024672 -0.052030   -0.050089  0.645993 -0.004068   
Education           0.021463  0.041334    0.013152 -0.187524 -0.017377   
Mortgage           -0.013920 -0.012539   -0.010582  0.206806  0.007383   
Personal Loan      -0.024801 -0.007726   -0.007413  0.502462  0.000107   
Securities Account -0.016972 -0.000436   -0.001232 -0.002616  0.004704   
CD Account         -0.006909  0.008043    0.010353  0.169738  0.019972   
Online             -0.002528  0.013702    0.013898  0.014206  0.016990   
CreditCard          0.017028  0.007681    0.008967 -0.002385  0.007691   

                      Family     CCAvg  Education  Mortgage  Personal Loan  \
ID                 -0.016797 -0.024672   0.021463 -0.013920      -0.024801   
Age                -0.046418 -0.052030   0.041334 -0.012539      -0.007726   
Experience         -0.052563 -0.050089   0.013152 -0.010582      -0.007413   
Income             -0.157501  0.645993  -0.187524  0.206806       0.502462   
ZIP Code            0.011778 -0.004068  -0.017377  0.007383       0.000107   
Family              1.000000 -0.109285   0.064929 -0.020445       0.061367   
CCAvg              -0.109285  1.000000  -0.136138  0.109909       0.366891   
Education           0.064929 -0.136138   1.000000 -0.033327       0.136722   
Mortgage           -0.020445  0.109909  -0.033327  1.000000       0.142095   
Personal Loan       0.061367  0.366891   0.136722  0.142095       1.000000   
Securities Account  0.019994  0.015087  -0.010812 -0.005411       0.021954   
CD Account          0.014110  0.136537   0.013934  0.089311       0.316355   
Online              0.010354 -0.003620  -0.015004 -0.005995       0.006278   
CreditCard          0.011588 -0.006686  -0.011014 -0.007231       0.002802   

                    Securities Account  CD Account    Online  CreditCard  
ID                           -0.016972   -0.006909 -0.002528    0.017028  
Age                          -0.000436    0.008043  0.013702    0.007681  
Experience                   -0.001232    0.010353  0.013898    0.008967  
Income                       -0.002616    0.169738  0.014206   -0.002385  
ZIP Code                      0.004704    0.019972  0.016990    0.007691  
Family                        0.019994    0.014110  0.010354    0.011588  
CCAvg                         0.015087    0.136537 -0.003620   -0.006686  
Education                    -0.010812    0.013934 -0.015004   -0.011014  
Mortgage                     -0.005411    0.089311 -0.005995   -0.007231  
Personal Loan                 0.021954    0.316355  0.006278    0.002802  
Securities Account            1.000000    0.317034  0.012627   -0.015028  
CD Account                    0.317034    1.000000  0.175880    0.278644  
Online                        0.012627    0.175880  1.000000    0.004210  
CreditCard                   -0.015028    0.278644  0.004210    1.000000

# drop useless feature
#Like Id, Zipcode

In [11]:
database2=database.drop(["ID","ZIP Code"],axis=1)

In [12]:
database2.head()

Age  Experience  Income  Family  CCAvg  Education  Mortgage  Personal Loan  \
0   25           1      49       4    1.6          1         0              0   
1   45          19      34       3    1.5          1         0              0   
2   39          15      11       1    1.0          1         0              0   
3   35           9     100       1    2.7          2         0              0   
4   35           8      45       4    1.0          2         0              0   

   Securities Account  CD Account  Online  CreditCard  
0                   1           0       0           0  
1                   1           0       0           0  
2                   0           0       0           0  
3                   0           0       0           0  
4                   0           0       0           1

# Convert (" Personal Loan") float value into Numerical

In [13]:
import numpy as np

In [14]:
database2["Personal Loan"]=np.round(database2["Personal Loan"])

In [15]:
database2["Personal Loan"] is float

False

In [16]:
database2.dtypes

Age                     int64
Experience              int64
Income                  int64
Family                  int64
CCAvg                 float64
Education               int64
Mortgage                int64
Personal Loan           int64
Securities Account      int64
CD Account              int64
Online                  int64
CreditCard              int64
dtype: object

In [17]:
database2["CCAvg"]=np.round(database2["CCAvg"])

In [18]:
database.dtypes

ID                      int64
Age                     int64
Experience              int64
Income                  int64
ZIP Code                int64
Family                  int64
CCAvg                 float64
Education               int64
Mortgage                int64
Personal Loan           int64
Securities Account      int64
CD Account              int64
Online                  int64
CreditCard              int64
dtype: object

In [19]:
database2["CCAvg"] is float

False

In [20]:
database2["CCAvg"]

0       2.0
1       2.0
2       1.0
3       3.0
4       1.0
       ... 
4995    2.0
4996    0.0
4997    0.0
4998    0.0
4999    1.0
Name: CCAvg, Length: 5000, dtype: float64

In [21]:
database2.head(5)

Age  Experience  Income  Family  CCAvg  Education  Mortgage  Personal Loan  \
0   25           1      49       4    2.0          1         0              0   
1   45          19      34       3    2.0          1         0              0   
2   39          15      11       1    1.0          1         0              0   
3   35           9     100       1    3.0          2         0              0   
4   35           8      45       4    1.0          2         0              0   

   Securities Account  CD Account  Online  CreditCard  
0                   1           0       0           0  
1                   1           0       0           0  
2                   0           0       0           0  
3                   0           0       0           0  
4                   0           0       0           1

# created Random forest Alog

In [22]:
from sklearn.ensemble import RandomForestClassifier

In [23]:
rf_model=RandomForestClassifier(n_estimators=1000,max_features=2,oob_score=True)

In [24]:
database2.columns

Index(['Age', 'Experience', 'Income', 'Family', 'CCAvg', 'Education',
       'Mortgage', 'Personal Loan', 'Securities Account', 'CD Account',
       'Online', 'CreditCard'],
      dtype='object')

In [25]:
feactures=['Age', 'Experience', 'Income', 'Family', 'CCAvg', 'Education',
       'Mortgage', 'Securities Account', 'CD Account',
       'Online', 'CreditCard']

In [26]:
rf_model.fit(X=database2[feactures],y=database2["Personal Loan"])

RandomForestClassifier(max_features=2, n_estimators=1000, oob_score=True)

In [28]:
print("OOB Accuary Score:")
print(rf_model.oob_score_)

OOB Accuary Score:
0.9862


In [31]:
for feacture ,imp in zip(feactures,rf_model.feature_importances_):
    print(feacture,imp)

Age 0.05006197947350431
Experience 0.050727004914156
Income 0.36731368800489456
Family 0.09890917133595348
CCAvg 0.14287863076258792
Education 0.15829381277475146
Mortgage 0.04867657205753933
Securities Account 0.006344575076331396
CD Account 0.056154270376598814
Online 0.009359036897629989
CreditCard 0.011281258326052972


# Random Forest Result

Income 0.36731368800489456

CCAvg 0.14287863076258792

Education 0.15829381277475146

# used Decision tress Alogrithms

In [32]:
from sklearn import tree

In [35]:
tree_model=tree.DecisionTreeClassifier(max_depth=6,max_leaf_nodes=10)

In [38]:
predicoter=pd.DataFrame([database2["Education"],database2["CCAvg"],database2["Income"]]).T

In [39]:
predicoter

Education  CCAvg  Income
0           1.0    2.0    49.0
1           1.0    2.0    34.0
2           1.0    1.0    11.0
3           2.0    3.0   100.0
4           2.0    1.0    45.0
...         ...    ...     ...
4995        3.0    2.0    40.0
4996        1.0    0.0    15.0
4997        3.0    0.0    24.0
4998        2.0    0.0    49.0
4999        1.0    1.0    83.0

[5000 rows x 3 columns]

In [40]:
tree_model.fit(X=predicoter,y=database2["Personal Loan"])

DecisionTreeClassifier(max_depth=6, max_leaf_nodes=10)

In [41]:
with open("Dtree.dot","w") as f:
    f=tree.export_graphviz(tree_model,feature_names=["Education","CCAvg","Income"],out_file=f)